### ISSUE TITLE
Consider disabling the Drupal compatibility layer when running tests

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#6037]
#6036 has brought to light a bug that only happens when the Drupal compatibility layer is disabled.

@herbdool 
> I'm thinking we should have an automated test which disables the compatibility layer to catch things like this. Or disable it for all tests.

**Comment by quicksketch** [comment]
I made a proof of concept PR at https://github.com/backdrop/backdrop/pull/4389 that simply disables the compatibility layer entirely. I don't think it will pass because I think we actually test the compatibility layer itself. But this should give us a starting point to see the failures.

**Comment by quicksketch** [comment]
The _only_ test that failed was the Entity Reference test, which should be fixed by https://github.com/backdrop/backdrop-issues/issues/6036.

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop-issues/issues/6036 and rebased the PR, we'll see if it all passes now.

It seems we don't have test coverage on the compatibility layer, which mostly just wraps functions and methods in new versions, but should we add testing, we'll need to find a way to toggle it on and off, perhaps by writing or updating the settings.local.php file like we're doing in the PR, but dynamically based on which tests we want to use it in.

**Comment by quicksketch** [comment]
Before #6036 with compatibility disabled (failing): https://github.com/backdrop/backdrop/actions/runs/4506265093
After #6036 with compatibility disabled (passing): https://github.com/backdrop/backdrop/actions/runs/4506457732

So this demonstrates that the PR turns off the compatibility mode, otherwise both testing runs would have passed.

**Comment by herbdool** [comment]
This looks good to me. We can probably leave the question of how to toggle it for another issue if we had tests for the layer itself.

**Comment by quicksketch** [comment]
Super, thanks @herbdool and @klonos! Merged https://github.com/backdrop/backdrop/pull/4389 into 1.x and 1.24.x.
